Delay skill rust for 24h after practicing

dseguin opened the pull request:
#### Summary
None

#### Purpose of change
Fixes #51647

#### Describe the solution
If `SkillLevel::_lastPracticed` less that 24 hours, don't calculate skill rust. This effectively delays skill rust until the next "tick", 24 hours after practicing the skill.

#### Describe alternatives you've considered
Previously, I was using a "grace point" system to delay skill rust. The issue with that was pointed out in the comments below.

#### Testing
1. Practice carving to level up fabrication to 1
2. Wait until the 24-hour cycle recalculates skill rust
3. Notice that it now takes 2 cycles to start applying skill rust to fabrication

#### Additional context
See the previous edit of this PR to read the original description.

BrettDong commented on src/skill.cpp:
```diff
@@ -257,15 +257,30 @@ void SkillLevel::train( int amount, float catchup_modifier, float knowledge_modi
+        float lvl_exp = lvl_mult * 10000.0f;
+        float rust_div = 0.04f;
+        if( _rustAccumulator < lvl_exp * 3 ) {
+            rust_div = std::max( static_cast<float>( std::sqrt( _rustAccumulator / lvl_exp ) ), 0.04f );
```
Is it possible that `_rustAccumulator < 0` at this moment?

dseguin replied:
I don't see why it would, but I'll add a condition just in case

BrettDong replied:
I think the NaN error is because a negative `_rustAccumulator` was passed in.

dseguin replied:
No problem, I moved the negative check to before this calculation. Hopefully that fixes it.

I-am-Erk commented:
It doesn't have to be in this pr, but it would be nice if some aspect of this were moderated by a global variable, allowing mods and individual settings to easily tweak it.

kevingranade reviewed (changes requested):
This has a lot more moving parts than just refraining from reporting on a practical skill shortfall until the magnitude is larger.

I-am-Erk commented:
That's true, but I think it's reasonable to postpone rusting a newly gained skill level since it leads to some counterintuitive behaviors. Is there something specific about this system you dislike?

I-am-Erk commented:
When I redid rust I skipped tracking "last practiced" because it was obnoxious but yes, that would be a better way to do this. What we could do is have "grace period" get reset to the value outlined in this PR when you practice rather than when you gain a level. That would still accomplish the goal of not dropping a newly gained level almost immediately, while not artificially enshrining level gain as a special case.

I also agree that the rust messages could be improved or simply turned off.

dseguin commented:
@I-am-Erk I implemented your suggestions to make this a bit simpler. I scrapped the "grace point" stuff in favor of `SkillLevel::_lastPracticed`. I also added a menu option in the debug category to tweak it more easily.

eltank commented on src/skill.cpp:
```diff
@@ -320,6 +321,12 @@ bool SkillLevel::isRusty() const
            _knowledgeExperience - _exercise >= ( _level + 1 ) * ( _level + 1 ) * 10;
 }
 
+bool SkillLevel::rust_grace_period() const
```
Does this need a separate function? Seems that you could simply put this code near the top of rust() below.
The name of the function also doesn't match the return value (it doesn't return a "period").

dseguin replied:
That's fair. I removed this function.

dseguin commented:
Superseded by #51724

dseguin commented:
Updated to remove menu option. The only thing this PR does is delay skill rust by 24 hours after practicing.
